### ISSUE TITLE
[Proposal] Collaborative RLVR for Robust Reasoning

### ISSUE THREAD

**Issue by swisski** [uchicago-nlp-course/winter-2026-students#20]
[Project proposal NLP.pdf](https://github.com/user-attachments/files/25170299/Project.proposal.NLP.pdf)

**Comment by qrdai** [comment]
## 1. strengths

1. an interesting research question that is presented as a nice combination of several prior works. solid and suitable for a course project.

2. i like how you anticipate the deliverable and fallbacks for each week. great research habits!

3. great literature search and presentation. i like the one-sentence summarization of each literature, and how it contributes to your proposal. it helps!


## 2. weaknesses

1. i'm not sure if a frozen coach makes enough sense for generating adaptive training data. in the original work [CPMobius], the coach seems to be continuously rewarded and updated during the training process, so as to keep up with the capability improvement of the player. to me, a frozen coach doesn't differ much from a fixed training dataset, and may be inferior since the training questions cannot go beyond the distribution of the initial coach. if you do want to enable adaptive curriculum for GRPO training data, try the techniques in DAPO (https://arxiv.org/abs/2503.14476). this is still the industrial standard for rlvr training as of Feb 2026. 

2. be prepared for problems with training efficiency and stability. multi-agent rlvr training is notorious for the latency in sampling long rollouts and the correspondingly sparse outcome rewards. i suggest checking the appendix of the main papers that you reference to make sure the training efficiency is acceptable (e.g., how many gpus and gpus hours they use). besides, it would be helpful to consider a few drop-in upgrades that stablize rlvr training, such as techniques in DAPO (other than the curriculum) and DR.GRPO (https://arxiv.org/abs/2503.20783). finally, not sure whether it would be helpful, but for multi-agent rl training, there are a few dedicated open-source projects that are under continuous maintenance, which may be better than the original repo that came along with MAPoRL (it was published in ACL 2025 so was mostly a work in the pre-deepseek-r1 era. the code may not be that optimal from today's perspective).
    - https://github.com/langfengQ/verl-agent
    - https://github.com/WentseChen/Verlog
    - https://verl.readthedocs.io/en/latest/sglang_multiturn/interaction_system.html


## 3. next steps
looks good to me. just remember to update at least one new figure / plot / table in each new weekly blog

**Comment by swisski** [comment]
[Week 7 Blog Post - Collaborative RLVR](https://github.com/user-attachments/files/25475689/Week_7_Blog_Post___RLVR_COLLAB-1.pdf)

**Comment by qrdai** [comment]
## 1. strengths

1. i like how your blog presents your step-by-step progress and takeaways. 
2. it's really amazing that you've achieved so much progress in only one week. keep it up!


## 2. weaknesses and next steps

1. **base models and benchmarks**. you have found that starting directly from qwen2.5-3b-base makes it hard to obtain meaningful gradient signals at the beginning. this is pretty much expected, since the base model hasn't even learned how to respond to an instruction appropriately (e.g., including the final answer in \boxed{}). since our innovations are focusing on rl alone (and we don't want to be distracted by sft), the mature approach in the community is to start with an instruct version of qwen (e.g., qwen2.5-3b-instruct, or qwen3-4b-nonthinking) or the math version of qwen (e.g., qwen2.5-7b-math-base; this model undergoes continued pretraining in an additional math corpora compared with the normal qwen2.5-7b-base). correspondingly, since the instruct/math versions of qwen already fare well on MATH/GSM8K, the community standard for math reasoning is now AIME24/25. therefore, i suggest switching to a stronger base model, and also targeting more difficult/modern math benchmarks, so as not to worry about the warmup sft.

2. **highly recommended: reading some tech reports to learn about the mature RLVR training recipe in the community**. it's awesome that you discovered the importance of a large rollout group size through your own experiments, but actually this has been a community consensus for a long time. a few other questions/observations you raised in the blog (e.g., the performance regression for the final checkpoints), including the point 1 above, also belong to such consensus. therefore, it would save you much time by reading the community practices first so as not to spend too much time on trial and error. a few tech reports and blogs i found useful include:
    - https://hkunlp.github.io/blog/2025/Polaris/
    - https://arxiv.org/abs/2505.22312
    - https://arxiv.org/abs/2512.13961
    - i also checked out the original contents of cpmobius and r-zero. they are both well written papers and i actually know the authors of both (esp. Chengsong, the first author of r-zero; if you do need more advice/prior experience about data-free rl engineering, and do want to target COLM with this submission, i can ask if Chengsong would be happy to advise this project). basically i recommend reading the technical setups of cpmobius (and r-zero), especially those embedded in the appendix, and discussing with chatgpt on your customized engineering decisions that are different from what cpmobius deploys.

3. on the **problem formulation and motivation** side, if we do want to eventually make it a submission for COLM, then we need to think more about **why it is necessary to incorporate both (1) using an adaptive coach as an adaptive data generation engine; (2) involving the discussion between two players.** i personally accept (1) better since it originates from the "data-free" constraints, and "data-free" is indeed a practical setup to consider. **but it's unclear to me what the unique or irreplaceable purpose of (2) is**. in the abstract of this blog, you mentioned "alleviating distribution gaming, unfaithful reasoning, and brittleness to rephrasings", but they are not convincing enough to me -- why **must** we use two-player discussion and invest so many engineering efforts? is it really true that these issues do exist in a significant manner without two-player discussion, and that two-player discussion can **more effectively** mitigate them than other more common and widely accepted rl engineering approaches?
    - for (1), i actually also feel that math may not be the domain where the adaptive coach works best, since it's pretty easy to obtain an abundance of high-quality math data. in what domains is data-free rl most valuable? this question may be too hard for such a course project, but if we do find a task that desparately needs data-free rl (i.e., desparately hard to scalably obtain high-quality instructions), then i bet we will make a super strong COLM submission.

4. **gpu resources**. try `modal shell` to get more gpu resources from Modal (https://modal.com/docs/reference/cli/shell). in this way, you may try full rlvr for both student models instead of using lora (for a single 3b model with ~8k response length, 4*80 GB should work for rl). but anyway, the key is to not get constrained by gpu resources by better leveraging the cloud resources that we have provided -- i highly appreciate your smart solution to save forward passes, but this may not deserve much of your time especially in the early stage of a project.

**feel free to lmk if you'd like to make this paper a COLM submission.** i think it has the potential, but we need to address a few fundamental issues in problem formulation, as well as take advice from some more experienced rl engineers. feel free to drop by my office hours to discuss more. **if you decide to aim for COLM, then we can schedule weekly/biweekly meetings accordingly:)**

**Comment by swisski** [comment]
[Collaborative RLVR Week 8 Blog](https://github.com/user-attachments/files/25673114/Collaborative.RLVR.Week.8.Blog.1.pdf)

**Comment by qrdai** [comment]
1. for player model, I still recommend trying the instruct mode (NOTE: not the thinking mode as it would usually generate extremely long chain-of-thoughts that would likely lead to OOM) of [`qwen3-4b`](https://huggingface.co/Qwen/Qwen3-4B), or the stronger [`qwen3-4b-2507`](https://huggingface.co/Qwen/Qwen3-4B-Instruct-2507), apart from the `qwen3-base`. from my observation about 30%/70% RL engineering works use pure base models / base models that undergo continued pretraining on math (e.g., `qwen2.5-math`), or instruct models. so it would be useful to try instruct as base models as well, and they can save you the trouble during the initial cold-start phase, makeing engineering easier.
    - i’d also like to clarify the differences between three versions of qwen models. let’s take the 4B versions of qwen3 as an example. `qwen3-4b-base`: the base model w/o post-training; ususally not equipped with instruction-following capabilities; `qwen3-4b`: the post-trained model based on `qwen3-4b-base`. it has two modes, instruct (enable_thinking=False) and thinking (enable_thinking=True). note that the instruct mode still outputs CoTs — it’s just shorter (usually <2K) and doesn’t involve frequent occurences of “Aha” or “Wait” (i.e., the deepseek-R1 behaviors, usually dubbed “long-CoT”); the thinking mode’s CoT is usually much longer (~16K). finally, `qwen3-4b-instruct-2507`: this is an upgraded version of `qwen3-4b` in the instruct mode. it can be seen as a drop-in replacement for `qwen3-4b` instruct mode, but note that only a few members in the qwen3 family have such upgrades. 


2. for coach reward design, i think it’s a good attempt to have coach solve the problem in thinking mode so as to avoid the player’s collapse in majority vote. one caveat is that the coach is rewarded and updated not based on its own problem-solving ability, but how well it generates a meaningful question for players. not sure if the coach would also get better in solving the question it generates through the updating process, but this is a great attempt after all. i’m curious to know how the “question design reward” would meanwhile improve the question solving capability of the coach.
    - glad that you are taking the “global $\Delta_t$ VS local instruction $i$” problem that i mentioned into careful consideration. though there might be global VS local conflicts, i think $\Delta_t$ may still be a valuable signal as it indicates a solid direction for capability improvement — downstream performance. A question can be considered as good only after it’s proved to be effective on downstream tasks; while all other reward components, such as uncertainty reward and discussion productivity bonus, are all indirect proxies for the downstream effectiveness of a question. implementation-wise, i suggest that we still start from the $\Delta_t$ formulation by CPMobius even though there might be global VS local conflicts. We can try adding new design components, such as the discussion bonus, onto the original design; or we can also try ablating the $\Delta_t$ term from the coach reward to see how it affects the final performance for both players and the coach.
    - the repetition penalty is a great design! glad that you noticed this missing piece from CPMobius. embedding-based deduplication looks sound to me, though we may also try comparing with R-Zero’s BLEU-based penalty if time permits.


3. for the combination of (player model, training dataset for condition A&C), I think MATH, even though level 3-5, may still be a bit easy for Qwen3-4B (both base and instruct) models, especially their instruct variants. From the Qwen3 technical report, Qwen3-4B-base achieves ~54% on MATH and Qwen3-4B (instruct mode) achieves ~85%. Maybe the difficult data inside [DAPO-17K](https://huggingface.co/datasets/BytedTsinghua-SIA/DAPO-Math-17k) could work better? I’m not sure, but feel free to try out MATH level 3-5 first.


4. for other updates on algorithmw-wise design choices, I think they sound good, but it may be safer if we start from a combination of design choices that is already extensively tested. for example, I think we can start with the “clip-high” and “reverse-KL” configurations in the standard DAPO, and try both MAGPRO and Dr. MAS at the beginning, before we decide to update the clipping or divergence strategy. On one hand, this again enables us to sanity check the numbers against those reported in DAPO/CPMobius, and makes our life easier when it comes to debugging. On the other hand, even though a paper proposes new RL engineering techniques and experimentally shows it works, it may not really work well, especially when generalizing to a different setup from the original paper (actually, the effectiveness RL engineering techniques are notoriously hard to replicate in most cases, even under exactly the same setup as the original paper). The community usually leans more towards extensively tested techniques such as DAPO/DR. GRPO, or the comprehensive technical reports released by renouned industry labs (e.g., the Skywork/OMLO3/HKUNLP tech reports I shared in last week’s feedback)


5. about the smoke test results: would be great if more technical details about the eval setup can be revealed. the current curves look a bit too magic to me.
    - what dataset(s) is the accuracy computed on? the full AIME24/25 suite?
    - according to the Qwen3 technical reports, the performance of qwen3-4b-base on gsm8k already hits 87.8%, which seems contradictory to the 75% diagreement rate reported after the first iteration. 
    - how is the evaluation carried out for multi-players? say after the first iteration, do you take both players at this stage, feed the AIME dataset to them, still allow them to discuss, and eventually return a consensus answer as the final answer? the current trends seem a bit too good to be true with only 2-3 iterations (not sure if you do mean “iterations” or  “epochs”) of training, so I recommend looking into the detailed logs of training and eval to make sure everything is working correctly.

**Comment by swisski** [comment]
Final paper: 

[CRISP.pdf](https://github.com/user-attachments/files/25960260/CRISP.pdf)
